pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +302,473 @@ pub mod pallet {
-		/// having increased their investment as this would store the payment
-		/// currency.
-		InvestmentPaymentCurrencyNotFound,
-		/// Failed to retrieve the foreign payout currency for a collected
-		/// redemption.
-		///
-		/// NOTE: This error can only occur, if a user tries to collect before
-		/// having increased their redemption as this would store the payout
-		/// currency.
-		RedemptionPayoutCurrencyNotFound,
-		/// Failed to retrieve the `TokenSwapReason` from the given
-		/// `TokenSwapOrderId`.
-		InvestmentInfoNotFound,
-		/// Failed to retrieve the `TokenSwapReason` from the given
-		/// `TokenSwapOrderId`.
-		TokenSwapReasonNotFound,
-		/// The fulfilled token swap amount exceeds the sum of active swap
-		/// amounts of the corresponding `InvestmentState` and
-		/// `RedemptionState`.
-		FulfilledTokenSwapAmountOverflow,
-		/// Failed to transition the `InvestState`.
-		InvestError(InvestError),
-		/// Failed to transition the `RedeemState.`
-		RedeemError(RedeemError),
+		/// Failed to retrieve the `ForeignInvestInfo`.
+		InfoNotFound,
+		/// Failed to retrieve the swap order.
+		SwapOrderNotFound,
 		/// Failed to retrieve the pool for the given pool id.
 		PoolNotFound,
 	}
+
+	/// Internal type used as result of `Pallet::apply_swap()`
+	struct SwapStatus<T: Config> {
+		/// The amount pending to be swapped
+		pending: T::Balance,
+
+		/// The amount already swapped and available to use
+		swapped: T::Balance,
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn register_swap(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			kind: ForeignSwapKind,
+			swap_id: T::SwapId,
+		) {
+			ForeignSwapIdToSwapId::<T>::insert(&who, (investment_id, kind), swap_id);
+			SwapidToForeignSwapId::<T>::insert(swap_id, (who.clone(), investment_id, kind));
+		}
+
+		fn unregister_swap(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			kind: ForeignSwapKind,
+			swap_id: T::SwapId,
+		) {
+			ForeignSwapIdToSwapId::<T>::remove(&who, (investment_id, kind));
+			SwapidToForeignSwapId::<T>::remove(swap_id);
+		}
+
+		/// Returns the `amount_out` of the swap, seeing as the amount used to
+		/// swap.
+		fn used_swap_amount(swap: &SwapOf<T>) -> Result<T::Balance, DispatchError> {
+			T::CurrencyConverter::stable_to_stable(
+				swap.currency_out,
+				swap.currency_in,
+				swap.amount_in,
+			)
+		}
+
+		/// Apply a swap over a current possible swap state.
+		/// - If there was no previous swap, it adds it.
+		/// - If there was a swap in the same direction, it increments it.
+		/// - If there was a swap in the opposite direction:
+		///   - If the amount is smaller, it decrements it.
+		///   - If the amount is the same, it removes the inverse swap.
+		///   - If the amount is greater, it removes the inverse swap and create
+		///     another with the excess
+		///
+		/// The returned status contains the swapped amount after this call and
+		/// the pending amount to be swapped.
+		fn apply_swap(
```

Comment:
This is the key to the refactoring. All swapping processes happen here, independently of the direction or if it's an investment or redemption. Basically, this method extracts the swapping logic from `InvestState` and `RedeemState`. Later, the increase/decrease method uses `apply_swap()` to perform their needs.

It only uses the `TokenSwaps` trait, so mocking this effect should be enough to test all possible swap combinations isolated.

Reply:
Ideally, after https://github.com/centrifuge/centrifuge-chain/pull/1674, we only need to touch this method to update the FI to support new the order book implementation with `amount_out` instead of `amount_in`